src/libraries/System.Net.Http/src/System/Net/Http/HttpResponseMessage.cs

```diff
@@ -22,6 +22,8 @@ public class HttpResponseMessage : IDisposable
         private HttpContent? _content;
         private bool _disposed;
 
+        internal IHttpMetricsLogger? HttpMetricsLogger { get; set; }
```

Comment:
We would need to wrap `HttpContent` for that, which would add complexity. `HttpResponseMessage` together with it's `HttpContent` is a heavy object, so adding a field makes little difference IMO.

Reply:
> so adding a field makes little difference IMO

There's no need for opinions here... it can be measured.

In principle, though, it'd be nice to make things pay-for-play, and this is making every HttpResponseMessage larger, regardless of whether metrics are enabled, and regardless of whether you're even using SocketsHttpHandler.

> We would need to wrap HttpContent for that

Are those the only two options?  For example, the response references the request, and the request has the options bag; if it's likely it will have been created in support of metrics and enrichment, could this object just be stored into that collection?  Or we've chosen to make "MetricsHandler" an implementation detail within SocketsHttpHandler, so we can double-down on it being an implementation detail and have the actual message that's produced by the core implementation be a derived type with more stuff on it.  Or add an internal virtual method to HttpMessageHandler to be called by HttpClient when a response fails; SocketsHttpHandler would just delegate that through its handler chain to MetricsHandler. Etc.

It's possible the extra 8 bytes per request simply doesn't matter.  I just don't like the idea of us making this object more expensive for everyone in support of a failure condition for optional metrics on one specific implementation of the abstraction.

Reply:
> Or add an internal virtual method to HttpMessageHandler to be called by HttpClient when a response fails; SocketsHttpHandler would just delegate that through its handler chain to MetricsHandler

We can't enforce that an arbitrary handler chain would delegate to SocketsHttpHandler.

>  the request has the options bag; if it's likely it will have been created in support of metrics and enrichment, could this object just be stored into that collection?

Currently the collection is only created for enrichment-enabled requests, which is only a subset of metrics-enabled requests. To not penalize metrics usage that doesn't do enrichment, we would need to commit to pool the option bag objects.

In fact I think it should be possible to only pool that thing, and not the `HttpMetricsEnrichmentContext`. Will do a bigger refactor to see how that would look like.

Reply:
> We can't enforce that an arbitrary handler chain would delegate to SocketsHttpHandler.

I think I'm missing something. I thought we'd agreed that we'd track until headers were received, not until the whole response stream is processed. Is that not the case? If we conditionally track until EOF only when ResponseContentRead is set, then we're going to have a mishmash of metrics, some for just headers, some for content. But if we limit it to anyways just being when response headers are received, then there's no concern about handler chains; everything is internal to the terminating handler, eg SocketsHttpHandler. When I referred to adding a virtual methods to the handler, i meant to the handler stage inside of SocketsHttpHandler, which MetricsHandler would be.

Reply:
> I thought we'd agreed that we'd track until headers were received, not until the whole response stream is processed. Is that not the case?

I believe we made that decision for the duration measurement, but I don't recall if the discussion ever explicitly stated whether the decision also applied to other metrics. I assumed it would apply to all the http metrics so I'm with Stephen in believing this extra tracking is unnecessary.

Reply:
I want to match the behavior of the existing `request-failed` DiagnosticCounter and the `Request/Failed` telemetry event. Both of them *do* record errors which occur during buffering in `HttpClient` when `HttpCompletionOptions.ResponseContentRead` is being passed, but they don't bother to do the same for the exceptions thrown if the response stream is being read by the user after passing `HttpCompletionOptions.HeadersReceived`. Unfortunately, there is no one around right now who I could ask why did we decide to go with this behavior back in the days, but I assume the decision was thought through. I don't want the metrics counter `http-client-failed-requests` to be "weaker", ie. not reporting errors in cases where it is recorded with DiagnosticCounter. I also don't see why should it be "stronger" (ie. wrap `HttpContent` and `HttpContentStream` to record all the errors) without a proof that users need it. I have sent an email earlier regarding this counter and R9 folks confirmed they don't care about it.

I'm using this property so we can record the metrics counter in our common error handling method in `HttpClient`:
https://github.com/dotnet/runtime/blob/1d43c3eca98af2a1fa0aa8b80542fc40bbcf859c/src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs#L624-L625

Reply:
Ah I see, we have two precedents and they are pulling us in opposite directions. The prior behavior of the EventCounters `failed-requests` counter encourages us to include these failures and our scoping of the duration metric encourages us not to include it. At best we can only be consistent with one of them and we have to decide which.

On the implementation side, I think there are alternatives that don't require any new fields on the HttpResponse. For example I think this works?
1. MessageHandler implementations that create the MetricsHandler cache it in a local field (per-chain storage rather than per-message)
2. Define an internal interface for those MessageHandlers to implement that has a LogRequestFailed() method on it. 
3. Implement the interface using the MetricsHandler field saved in (1)
4. HttpClient can attempt to cast its handler to that private interface and call the method rather than fetching it through the response.
This doesn't eliminate all perf overheads, but it does eliminate allocating and initializing the per-message field.

Are there some common error conditions that would generate Exceptions after the headers were parsed but before HttpClient finishes? If its a very noticeable difference and we aren't getting complaints about current EventCounters behavior I'd lean towards mirroring behavior we shipped before. If these error cases are fairly uncommon then I lean towards lets make the metrics more self-consistent and save a tiny bit of perf and implementation complexity.

Reply:
When thinking about implementation techniques, we should take into account that there might be any arbitrary HttpMessageHandler-s in the chain, we don't have ways to forward any queries to the innermost HttpClientHandler/SocketsHttpHandler. (We can only guarantee that if all of them are DelegatingHandlers with an InnerHandler reference.)

Regardless, there are many techniques we can choose from, and IMHO the implementation should not be the concern. I want us to agree on the functionality before talking about the implementation.

> Are there some common error conditions that would generate Exceptions after the headers were parsed but before HttpClient finishes

Unfortunately, I don't have any better data here other than anecdotally stating that "we have seen such errors occur for customers", but I think we should count with them and mirror the behavior we shipped with event counters.

@stephentoub thoughts?

Reply:
Actually one argument for matching the `request-duration` metrics counter instead of the event counter if  exposing `˛MetricsHandler` as a public independent component is something we want to count with for the future. In that case a hacky implementation that violates layering by using internals would be a problem.

Reply:
If you are uncertain about future maintainability I'd say go ahead and make the behavior change. Worst case we'll get some feedback from users that the behavior isn't what they want and then we can respond to it. I'm making an educated guess that feedback won't be that large if there is any at all.